#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: clock_{get,set}time@@GLIBC_2.0 - programs in Woody need it.
# DP:		   Add fix that the previous version does not provide @GLIBC_2.2 symbols.
# DP: Related bugs: http://lists.debian.org/debian-glibc/2004/12/msg00214.html
# DP: Dpatch author: GOTO Masanori <gotom@debian.org>
# DP: Patch author: Guido Guenther <agx@debian.org>, GOTO Masanori <gotom@debian.org>
# DP: Upstream status: Debian-Specific
# DP: Status Details: debian-specific, until sarge+1 will be released.
# DP: Date: 2004-12-24 (updated)

PATCHLEVEL=0

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
--- sysdeps/unix/sysv/linux/mips/Versions.orig	Tue Sep 24 15:56:55 2002
+++ sysdeps/unix/sysv/linux/mips/Versions	Tue Sep 24 15:49:27 2002
@@ -34,3 +34,9 @@
     _sys_errlist; sys_errlist; _sys_nerr; sys_nerr;
   }
 }
+librt {
+  GLIBC_2.0 {
+    # c*
+    clock_gettime; clock_settime;
+  }
+}
--- Versions.def	2004-12-21 00:37:04.000000000 +0900
+++ Versions.def.new	2004-12-21 09:50:31.000000000 +0900
@@ -82,6 +82,7 @@
   GLIBC_PRIVATE
 }
 librt {
+  GLIBC_2.0
   GLIBC_2.1
   GLIBC_2.2
   GLIBC_2.3
--- sysdeps/unix/clock_gettime.c~	2004-12-04 17:05:53.000000000 +0900
+++ sysdeps/unix/clock_gettime.c	2004-12-23 12:04:01.000000000 +0900
@@ -39,8 +39,13 @@
 
 
 /* Get current value of CLOCK and store it in TP.  */
+#ifdef DEBIAN_LIBRT_MIPS
+int
+__clock_gettime (clockid_t clock_id, struct timespec *tp)
+#else
 int
 clock_gettime (clockid_t clock_id, struct timespec *tp)
+#endif
 {
   int retval = -1;
 
--- sysdeps/unix/clock_settime.c~	2004-12-04 17:05:53.000000000 +0900
+++ sysdeps/unix/clock_settime.c	2004-12-23 12:09:27.000000000 +0900
@@ -37,8 +37,13 @@
 
 
 /* Set CLOCK to value TP.  */
+#ifdef DEBIAN_LIBRT_MIPS
+int
+__clock_settime (clockid_t clock_id, const struct timespec *tp)
+#else
 int
 clock_settime (clockid_t clock_id, const struct timespec *tp)
+#endif
 {
   int retval;
 
--- /dev/null	2004-09-27 19:55:25.000000000 +0900
+++ sysdeps/unix/sysv/linux/mips/clock_settime.c	2004-12-23 12:07:36.000000000 +0900
@@ -0,0 +1,18 @@
+/*
+ * It adds for debian specific librt-mips binary incompatibility problems.
+ * - gotom 2004-12-20.
+ */
+
+#define DEBIAN_LIBRT_MIPS
+
+/* include linux clock_settime */
+#include <sysdeps/unix/sysv/linux/clock_settime.c>
+
+/* compat symbols. */
+#include <shlib-compat.h>
+
+#if SHLIB_COMPAT (librt, GLIBC_2_0, GLIBC_2_2)
+strong_alias (__clock_settime, __clock_settime_old);
+compat_symbol (librt, __clock_settime_old, clock_settime, GLIBC_2_0);
+#endif
+versioned_symbol (librt, __clock_settime, clock_settime, GLIBC_2_2);
--- /dev/null	2004-09-27 19:55:25.000000000 +0900
+++ sysdeps/unix/sysv/linux/mips/clock_gettime.c	2004-12-23 12:07:35.000000000 +0900
@@ -0,0 +1,18 @@
+/*
+ * It adds for debian specific librt-mips binary incompatibility problems.
+ * - gotom 2004-12-20.
+ */
+
+#define DEBIAN_LIBRT_MIPS
+
+/* include linux clock_gettime */
+#include <sysdeps/unix/sysv/linux/clock_gettime.c>
+
+/* compat symbols. */
+#include <shlib-compat.h>
+
+#if SHLIB_COMPAT (librt, GLIBC_2_0, GLIBC_2_2)
+strong_alias (__clock_gettime, __clock_gettime_old);
+compat_symbol (librt, __clock_gettime_old, clock_gettime, GLIBC_2_0);
+#endif
+versioned_symbol (librt, __clock_gettime, clock_gettime, GLIBC_2_2);
